connectivity-check: Add DNS query tool for debugging

joestringer opened the pull request:
Add a new section to the connectivity checks specifically for manual
debugging. These new tools will not fail by themselves so they are not
appropriate for continuous integration testing.

The new tool can be generated specifically via:

    $ cue cmd -t type=tool -t component=proxy dump

Or by building the new generic connectivity-debug-tools.yaml target:

    $ make -C examples/kubernetes/connectivity-check connectivity-debug-tools.yaml

To use the new tool, deploy it and follow the logs:

    $ kubectl logs -l name=query-dns-policy --timestamps -f

Then take manual actions that may influence DNS proxy forwarding &
observe how the output of the pod changes over time.

Related: #12718

joestringer committed:
connectivity-check: Add DNS query tool for debugging

Add a new section to the connectivity checks specifically for manual
debugging. These new tools will not fail by themselves so they are not
appropriate for continuous integration testing.

The new tool can be generated specifically via:

 $ cue cmd -t type=tool -t component=proxy dump

Or by building the new generic connectivity-debug-tools.yaml target:

 $ make -C examples/kubernetes/connectivity-check connectivity-debug-tools.yaml

To use the new tool, deploy it and follow the logs:

 $ kubectl logs -l name=query-dns-policy --timestamps -f

Then take manual actions that may influence DNS proxy forwarding &
observe how the output of the pod changes over time.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on examples/kubernetes/connectivity-check/main_tool.cue:
```diff
@@ -15,14 +15,15 @@ objectSets: [
 ccCommand: {
 	#flags: {
-		component:  "all" | *"default" | "network" | "policy" | "services" | "hostport" | "proxy" @tag(component,short=all|default|network|policy|services|hostport|proxy)
+		component:  "all" | *"default" | "dns" | "network" | "policy" | "services" | "hostport" | "proxy" @tag(component,short=all|default|dns|network|policy|services|hostport|proxy)
```
This hunk is actually no longer necessary, I decided to put the new tool under "component: proxy" instead.

joestringer commented on examples/kubernetes/connectivity-check/tools.cue:
```diff
@@ -0,0 +1,34 @@
+// liveness/readiness probes are bypassed so it will always report success!
+deployment: "query-dns-policy": {
+	_image: "alpine"
+	_command: ["/bin/sh", "-c", "apk update && apk add bind-tools && while true; do dig +noall +question +answer +timeout=1 +tries=1 www.google.com && sleep 1 ; done"]
```
If anyone has suggestions for a simple container image that includes `dig`, we could substitute that here & drop the install step + CNP fqdn policy.

sayboras replied:
just curious if nslookup can be used here as well, then we can avoid apk commands, and remove related alpine egress policy.

Alternatively, I observed that we are using one image called alpine-curl in connectivity test, any plan to naintain one alpine image with (bind-tools + netutils + curl) for debugging purpose like what you have here as well ?

joestringer replied:
`nslookup` doesn't have the features we need for this test.

Looks like we can use `cilium/demo-client` docker image instead, it already has all the tools we need.

sayboras commented on examples/kubernetes/connectivity-check/Makefile:
```diff
@@ -48,6 +49,11 @@ $(QUARANTINE_OUT): $(SRC) $(SERVERS_OUT)
 
+$(TOOLS_OUT): $(SRC) $(SERVERS_OUT)
+	@echo '# Automatically generated by Makefile. DO NOT EDIT' > $(TOOLS_OUT)
+	$(QUIET)$(CUE) -t component=all -t type=tool dump > $@
```
as highlighted by @fristonio in another PR, >> should be used to avoid overriding first line.

joestringer committed:
connectivity-check: Add tool type[=autocheck]

Add a new label, "type" which will have two settings, either autocheck
(ie connectivity check script that fails based on certain conditions,
suitable for automated runtime usage in CI), and "tool" which requires
manual deployment & observation from a human.

Signed-off-by: Joe Stringer <joe@cilium.io>
